discovery/dns/dns.go

```diff
@@ -169,39 +170,117 @@ func (d *Discovery) refresh(ctx context.Context, name string, ch chan<- []*confi
+// Try and get an answer for various permutations of the given name,
+// appending the system-configured search path as necessary.
+//
+// There are three possible outcomes:
+//
+// 1. One of the permutations of the given name is recognised as
+//    "valid" by the DNS, in which case we consider ourselves "done"
+//    and that answer is returned.  Note that, due to the way the DNS
+//    handles "name has resource records, but none of the specified type",
+//    the answer received may have an empty set of results.
+//
+// 2.  All of the permutations of the given name are responded to by one of
+//    the servers in the "nameservers" list with the answer "that name does
+//    not exist" (NXDOMAIN).  In that case, it can be considered
+//    pseudo-authoritative that there are no records for that name.
+//
+// 3.  One or more of the names was responded to by all servers with some
+//    sort of error indication.  In that case, we can't know if, in fact,
+//    there are records for the name or not, so whatever state the
+//    configuration is in, we should keep it that way until we know for
+//    sure (by, presumably, all the names getting answers in the future).
+//
+// Outcome 1 is indicated by a valid response message and no error.  Outcome
+// 2 is a nil response message no error.  Outcome 3 is indicated by an error
+// return.  The error will be generic-looking, because trying to return all
+// the errors returned by the combination of all name permutations and
+// servers is a nightmare.
+//
+func lookupWithSearchPath(name string, qtype uint16, logger log.Logger) (*dns.Msg, error) {
 	conf, err := dns.ClientConfigFromFile(resolvConf)
 	if err != nil {
 		return nil, fmt.Errorf("could not load resolv.conf: %s", err)
 	}
 
-	client := &dns.Client{}
 	response := &dns.Msg{}
+	allResponsesValid := true
+
+	for _, lname := range conf.NameList(name) {
+		response, err = lookupName(lname, qtype, conf, logger)
+
+		if err != nil {
+			// We can't go home yet, because a later name
+			// may give us a valid, successful answer.  However
+			// we can no longer say "this name definitely doesn't
+			// exist", because we did not get that answer for
+			// at least one name.
+			allResponsesValid = false
+		} else if response.Rcode == dns.RcodeSuccess {
+			// Outcome 1: GOLD!
+			return response, nil
+		}
+	}
+
+	if allResponsesValid {
+		// Outcome 2: everyone says NXDOMAIN, that's good enough for me
+		return nil, nil
+	}
+	// Outcome 3: boned.
+	return nil, fmt.Errorf("could not resolve %s: one or more search-path-mangled names returned all errors", name)
+}
+
+// Lookup a specific name across all configured servers.  If a viable answer
+// is received from a server, then it is immediately returned, otherwise the
+// other servers in the config are tried, and if none of them return a
+// viable answer, an error is returned.
+//
+// A "viable answer" is one which indicates either:
```

Comment:
I appreciate that the expressive documentation has helped to form the mental model, but these comments are very long without providing additional information. The following would convey the same information `A viable answer has the rcode NoError or NXDomain. A non-viable answer is anything else, ...`.

Reply:
Since I'm not planning on maintaining this code, I don't really have a stake in whether the comments stay or go.  But after a long time of working with DNS, and helping other people figure out DNS, and looking at how other people have failed to figure out DNS, it feels like *very* few people actually understand how DNS works, and having as much clarity (and, yes, repetition, because nobody reads any more than they absolutely have to, so you need to repeat useful information in every single place someone might read it, and not assume they've absorbed everything) will, I'm pretty sure, help avoid nightmare scenarios in future.

Reply:
I question the repetition bit and that it makes docs more approachable. In my experience, if docs are too long, they will be just skipped completely.